src/data/complex/circle.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import geometry.manifold.instances.sphere
+
+/-!
+# The circle
+
+This file defines `circle` to be the metric sphere (`metric.sphere`) in `ℂ` centred at `0` of
+radius `1`.  We equip it with the following structure:
+
+* a submonoid of `ℂ`
+* a group
+* a topological group
+* a charted space with model space `euclidean_space ℝ (fin 1)` (inherited from `metric.sphere`)
+* a Lie group with model space `𝓡 1`
+
+We furthermore define `exp_map_circle` to be the natural map `λ t, exp (t * I)` from `ℝ` to
+`circle`, and show that this map is a group homomorphism and is smooth.
+
+## Implementation notes
+
+To borrow the smooth manifold structure cleanly, the circle needs to be defined using
+`metric.sphere`, and therefore the underlying set is `{z : ℂ | abs (z - 0) = 1}`.  This prevents
+certain algebraic facts from working definitionally -- for example, the circle is not defeq to
+`{z : ℂ | abs z = 1}`, which is the kernel of `complex.abs` considered as a homomorphism from `ℂ`
+to `ℝ`, nor is it defeq to `{z : ℂ | norm_sq z = 1}`, which is the kernel of the homomorphism
+`complex.norm_sq` from `ℂ` to `ℝ`.
+
+-/
+
+noncomputable theory
+
+open complex finite_dimensional metric
+open_locale manifold
+
+local attribute [instance] findim_real_complex_fact
+
+/-- The unit circle in `ℂ`, here given the structure of a submonoid of `ℂ`. -/
+def circle : submonoid ℂ :=
+{ carrier := sphere (0:ℂ) 1,
+  one_mem' := by simp,
+  mul_mem' := λ a b, begin
+    simp only [norm_eq_abs, mem_sphere_zero_iff_norm],
+    intros ha hb,
+    simp [ha, hb],
+  end }
+
+@[simp] lemma mem_circle_iff_abs (z : ℂ) : z ∈ circle ↔ abs z = 1 := mem_sphere_zero_iff_norm
+
+lemma circle_def : ↑circle = {z : ℂ | abs z = 1} := by { ext, simp }
+
+@[simp] lemma abs_eq_of_mem_circle (z : circle) : abs z = 1 := by { convert z.2, simp }
+
+instance : group circle :=
+{ inv := λ z, ⟨conj z, by simp⟩,
+  div := λ z, λ w, z * ⟨conj w, by simp⟩,
+  div_eq_mul_inv := by simp,
```

Comment:
`div_eq_mul_inv := λ z w, rfl` would have worked here too, right? If so, then you can just let these be generated automatically:
```suggestion
```

Reply:
Thanks!